{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Nadavo6/BioInformaticsProject/blob/main/mini_project_part_1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Proccessing the Database - Part 1\n",
        "(Habitat, Taxa and COG files)\n",
        "\n",
        "> Indented block\n",
        "\n"
      ],
      "metadata": {
        "id": "IosWtpgWpY8-"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "s8fLj2QvOsN0"
      },
      "outputs": [],
      "source": [
        "# Run this first to Create Data Dictionaries (Habitat, Taxa, Cog_info)\n",
        "\n",
        "# Prelimanary Code\n",
        "!pip install biopython\n",
        "from Bio import SeqIO\n",
        "from datetime import datetime\n",
        "\n",
        "# Creating Taxa Dictionary\n",
        "def process_taxa_file():\n",
        "    data_dict = {}\n",
        "    with open(taxa_file_name, 'r') as file:\n",
        "        for line in file:\n",
        "            parts = line.strip().split(',')\n",
        "            strain_id = parts[-3]\n",
        "            bacteria_id = parts[-4]\n",
        "            phylum_name = parts[1]\n",
        "            list = [strain_id, phylum_name]\n",
        "            if (strain_id!='strain_id'):\n",
        "              data_dict[bacteria_id] = list\n",
        "    return data_dict\n",
        "\n",
        "# # Creating cog_info Dictionary\n",
        "def process_cog_info_file(file_name, encoding='utf-8'):\n",
        "    data_dict = {}\n",
        "    with open(file_name, 'r', encoding=encoding, errors='replace') as file:\n",
        "        for line_number, line in enumerate(file, start=1):\n",
        "            try:\n",
        "                parts = line.strip().split(';')\n",
        "                if len(parts) == 6:\n",
        "                    cog = parts[0]\n",
        "                    key = cog[3:]\n",
        "                    value = parts[1]+';'+parts[2]+';'+parts[3]+';'+parts[4]\n",
        "                    data_dict[key] = value\n",
        "\n",
        "            except UnicodeDecodeError as e:\n",
        "                print(f\"Error decoding line {line_number}: {e}\")\n",
        "                print(f\"Problematic line content: {line}\")\n",
        "    return data_dict\n",
        "\n",
        "\n",
        "\n",
        "# Creating Habitat Dictionary\n",
        "def process_habitat_file():\n",
        "    data_dict = {}\n",
        "    with open(habitat_file_name, 'r') as file:\n",
        "        for line in file:\n",
        "            parts = line.strip().split(';')\n",
        "            host_name = parts[1]\n",
        "            habitat = parts[-1]\n",
        "            if (host_name!='bacteria'):\n",
        "              data_dict[host_name] = habitat\n",
        "    return data_dict\n",
        "\n",
        "\n",
        "# Main Code\n",
        "\n",
        "taxa_file_name = 'taxa.txt'\n",
        "taxa_dict = process_taxa_file()\n",
        "\n",
        "habitat_file_name = 'habitat.txt'\n",
        "habitat_dict = process_habitat_file()\n",
        "\n",
        "cog_file_name = 'COG_INFO_TABLE.txt'\n",
        "cog_dict = process_cog_info_file(cog_file_name)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Extracring Plasmid information - Part 1"
      ],
      "metadata": {
        "id": "QpIDOxzApo7L"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OOZaUvPmBU3Q",
        "collapsed": true
      },
      "outputs": [],
      "source": [
        "class Genome:\n",
        "    def __init__(self, Genome_id, creature_name):\n",
        "        self.Genome_id = Genome_id\n",
        "        self.creature_name = creature_name\n",
        "        self.segments = []\n",
        "\n",
        "def process_file(filename):\n",
        "    Genomes = {}\n",
        "    with open(filename, 'r') as file:\n",
        "        for line in file:\n",
        "            parts = line.strip().split('#')\n",
        "            creature_name = parts[3]\n",
        "            Genome_id = parts[1]\n",
        "            if Genome_id not in Genomes:\n",
        "                Genomes[Genome_id] = Genome(Genome_id, creature_name)\n",
        "            genome = Genomes[Genome_id]\n",
        "            segments = parts[-1].split()\n",
        "            current_segment = []\n",
        "            for segment in segments:\n",
        "                if segment.isdigit() and len(segment) == 4:\n",
        "                    current_segment.append(segment)\n",
        "                else:\n",
        "                    if current_segment:\n",
        "                        genome.segments.append(current_segment)\n",
        "                    current_segment = []\n",
        "            if current_segment:\n",
        "                genome.segments.append(current_segment)\n",
        "    return Genomes\n",
        "\n",
        "def get_clusters_of_length_d_at_least_q_times(Genomes, d, q):\n",
        "    segment_counts = {}\n",
        "\n",
        "    for Genome_id, Genome in Genomes.items():\n",
        "        for segment in Genome.segments:\n",
        "            if len(segment) >= d:\n",
        "                for i in range(len(segment) - d + 1):\n",
        "                    sub_segment = segment[i:i+d]\n",
        "                    segment_tuple = tuple(sorted(sub_segment))\n",
        "                    if segment_tuple in segment_counts:\n",
        "                        if Genome_id not in segment_counts[segment_tuple]:\n",
        "                            segment_counts[segment_tuple][Genome_id] = (sub_segment, Genome.creature_name)\n",
        "                    else:\n",
        "                        segment_counts[segment_tuple] = {Genome_id: (sub_segment, Genome.creature_name)}\n",
        "\n",
        "    filtered_clusters = {}\n",
        "    for segment, genomes in segment_counts.items():\n",
        "        if len(genomes) >= q:\n",
        "            filtered_clusters[segment] = genomes\n",
        "    top_clusters = dict(sorted(filtered_clusters.items(), key=lambda item: len(item[1]), reverse=True)[:5])\n",
        "    return top_clusters\n",
        "\n",
        "# Main code\n",
        "q = 4\n",
        "d = 2\n",
        "while (d<5):\n",
        "  filename = 'cog_words_plasmid.txt'\n",
        "  start_time = datetime.now()\n",
        "  Genomes = process_file(filename)\n",
        "\n",
        "  selected_lists = get_clusters_of_length_d_at_least_q_times(Genomes, d, q)\n",
        "  timedelta = datetime.now() - start_time\n",
        "  print(timedelta)\n",
        "  print()\n",
        "  print(\" \"*30 + \"Top 20 clusters of length\", d, \"appearing at least\", q, \"times:\")\n",
        "  print()\n",
        "  for segment, cluster_info in selected_lists.items():\n",
        "      print(\"Cluster:\", segment)\n",
        "      print(\"Occurrences:\", len(cluster_info))\n",
        "      print(\"Genome ID's and Creature Names:\")\n",
        "      for genome_id, (order, creature_name) in cluster_info.items():\n",
        "          print(f\"\\tGenome ID: {genome_id}, Creature Name: {creature_name}, Order: {order}\")\n",
        "      print(\"First Instance of this Cluster:\", next(iter(cluster_info.values()))[0])\n",
        "      print(\"Information about the COG's:\")\n",
        "      for cog in segment:\n",
        "        if cog in cog_dict:\n",
        "          print(\"\\t COG\",cog, \":\", cog_dict[cog])\n",
        "        else:\n",
        "          print(\"\\t COG\",cog, \":\", \"No Information was found for this COG\")\n",
        "      print(\"Statistics about the different orders we found:\")\n",
        "      word_counter = {}\n",
        "      host_list = {}\n",
        "      for value in cluster_info.values():\n",
        "        word_string = ', '.join(value[0])\n",
        "        word_counter[word_string] = word_counter.get(word_string, 0) + 1\n",
        "        host_name = value[1]\n",
        "        if word_string in host_list:\n",
        "          host_list[word_string].append(host_name)\n",
        "        else:\n",
        "          host_list[word_string]= []\n",
        "          host_list[word_string].append(host_name)\n",
        "      for key, value in word_counter.items():\n",
        "        print(\"\\t [\", key , \"]:\", value, \"times out of\", len(cluster_info), \"which is\", round(100*int(value)/len(cluster_info),2), \"percent\")\n",
        "      print(\"Statistics about the Phylum:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Phylum:\")\n",
        "        phylum_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_phylum = taxa_dict[host][1]\n",
        "          phylum_counter[curr_phylum] = phylum_counter.get(curr_phylum, 0) + 1\n",
        "        for key, value in phylum_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print(\"Statistics about the Habitat:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Habitats:\")\n",
        "        habitat_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_habitat = habitat_dict[host]\n",
        "          habitat_counter[curr_habitat] = habitat_counter.get(curr_habitat, 0) + 1\n",
        "        for key, value in habitat_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print()\n",
        "      print(\"*\"*100)\n",
        "  d = d+1\n",
        "  print(\"*\"*100)\n",
        "  print(\"*\"*100)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "  for segment, cluster_info in selected_lists.items():\n",
        "      print(\"Cluster:\", segment)\n",
        "      print(\"Occurrences:\", len(cluster_info))\n",
        "      print(\"Genome ID's and Creature Names:\")\n",
        "      for genome_id, (order, creature_name) in cluster_info.items():\n",
        "          print(f\"\\tGenome ID: {genome_id}, Creature Name: {creature_name}, Order: {order}\")\n",
        "      print(\"First Instance of this Cluster:\", next(iter(cluster_info.values()))[0])\n",
        "      print(\"Information about the COG's:\")\n",
        "      for cog in segment:\n",
        "        if cog in cog_dict:\n",
        "          print(\"\\t COG\",cog, \":\", cog_dict[cog])\n",
        "        else:\n",
        "          print(\"\\t COG\",cog, \":\", \"No Information was found for this COG\")\n",
        "      print(\"Statistics about the different orders we found:\")\n",
        "      word_counter = {}\n",
        "      host_list = {}\n",
        "      for value in cluster_info.values():\n",
        "        word_string = ', '.join(value[0])\n",
        "        word_counter[word_string] = word_counter.get(word_string, 0) + 1\n",
        "        host_name = value[1]\n",
        "        if word_string in host_list:\n",
        "          host_list[word_string].append(host_name)\n",
        "        else:\n",
        "          host_list[word_string]= []\n",
        "          host_list[word_string].append(host_name)\n",
        "      for key, value in word_counter.items():\n",
        "        print(\"\\t [\", key , \"]:\", value, \"times out of\", len(cluster_info), \"which is\", round(100*int(value)/len(cluster_info),2), \"percent\")\n",
        "      print(\"Statistics about the Phylum:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Phylum:\")\n",
        "        phylum_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_phylum = taxa_dict[host][1]\n",
        "          phylum_counter[curr_phylum] = phylum_counter.get(curr_phylum, 0) + 1\n",
        "        for key, value in phylum_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print(\"Statistics about the Habitat:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Habitats:\")\n",
        "        habitat_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_habitat = habitat_dict[host]\n",
        "          habitat_counter[curr_habitat] = habitat_counter.get(curr_habitat, 0) + 1\n",
        "        for key, value in habitat_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print()\n",
        "      print(\"*\"*100)\n",
        "  d = d+1\n",
        "  print(\"*\"*100)\n",
        "  print(\"*\"*100)"
      ],
      "metadata": {
        "id": "HdB0tSMEaNkN",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Filtering the Bac file - Part 2\n",
        "  (Here we clean the chromosomes we don't need from the bac file)\n",
        "\n",
        "> Indented block"
      ],
      "metadata": {
        "id": "Pkpzz7eTSnio"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Creating host_plasmid List\n",
        "# This is going over the plasmids, and making a list of every host for their's\n",
        "def process_plasmid_file(plasmid_file):\n",
        "    host_list = []\n",
        "    with open(plasmid_file, 'r') as file:\n",
        "        for line in file:\n",
        "          parts = line.strip().split('#')\n",
        "          host_name = parts[3]\n",
        "          if host_name not in host_list:\n",
        "            host_list.append(host_name)\n",
        "    return host_list\n",
        "\n",
        "plasmid_file_name = 'cog_words_plasmid.txt'\n",
        "plasmid_host_list = process_plasmid_file(plasmid_file_name)"
      ],
      "metadata": {
        "id": "9X0l2U1I09Ak"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# This Function returns the filtered bac file\n",
        "def copy_lines_with_condition(input, output, plasmid_search):\n",
        "    try:\n",
        "        # Open the bac file in read mode\n",
        "        with open(input, 'r') as input_file:\n",
        "            # Filter lines by plasmid_search function\n",
        "            bacs_that_have_plasmid = [line for line in input_file if plasmid_search(line)]\n",
        "        # Open a new bac file in write mode\n",
        "        with open(output_file_name, 'w') as output_file:\n",
        "            # Write the filtered lines to the new file\n",
        "            output_file.writelines(bacs_that_have_plasmid)\n",
        "        return output_file_name\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"The input file '{input_name}' does not exist.\")\n",
        "        return None\n",
        "\n",
        "# This Function recieves a line, extracts the host's name, and searches for it in plasmid_host list\n",
        "# returns true if Bac has plasmid, and False if not\n",
        "def plasmid_search(line):\n",
        "  chromo_has_plasmid = False\n",
        "  parts = line.strip().split('#')\n",
        "  host_name = parts[3]\n",
        "  for plasmid_host in plasmid_host_list:\n",
        "    if (plasmid_host == host_name):       ##This checks if this Chromosome's Host has at least one Plasmid\n",
        "      chromo_has_plasmid = True\n",
        "  if chromo_has_plasmid:\n",
        "    return True                   ## True = this host has already a plasmid in our DB\n",
        "  else:\n",
        "    return False                ## False = this host does not have a plasmid in our DB\n",
        "\n",
        "\n",
        "# The main Function\n",
        "input_file_name = \"cog_words_bac.txt\"\n",
        "output_file_name = \"new_cog_words_bac.txt\"\n",
        "result_file = copy_lines_with_condition(input_file_name, output_file_name, plasmid_search)\n",
        "\n",
        "# Downloading the filtered file - will appear in \"Downloads\"\n",
        "from google.colab import files\n",
        "files.download('new_cog_words_bac.txt')\n"
      ],
      "metadata": {
        "id": "5f3VxATr-_8n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Here you have to upload the new_cogs_words_bac file from your Downloads (21.1MB)!**"
      ],
      "metadata": {
        "id": "rppNQCiCj5xn"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VXgh4wT76qpO"
      },
      "source": [
        "# Extracring Chromosome information - Part 2"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# REMEMBER TO UPLOAD NEW_COG_WODS_BAC FILE!\n",
        "\n",
        "class Genome:\n",
        "    def __init__(self, Genome_id, creature_name):\n",
        "        self.Genome_id = Genome_id\n",
        "        self.creature_name = creature_name\n",
        "        self.segments = []\n",
        "\n",
        "def process_file(filename):\n",
        "    Genomes = {}\n",
        "    with open(filename, 'r') as file:\n",
        "        for line in file:\n",
        "            parts = line.strip().split('#')\n",
        "            creature_name = parts[3]\n",
        "            Genome_id = parts[1]\n",
        "            if Genome_id not in Genomes:\n",
        "                Genomes[Genome_id] = Genome(Genome_id, creature_name)\n",
        "            genome = Genomes[Genome_id]\n",
        "            segments = parts[-1].split()\n",
        "            current_segment = []\n",
        "            for segment in segments:\n",
        "                if segment.isdigit() and len(segment) == 4:\n",
        "                    current_segment.append(segment)\n",
        "                else:\n",
        "                    if current_segment:\n",
        "                        genome.segments.append(current_segment)\n",
        "                    current_segment = []\n",
        "            if current_segment:\n",
        "                genome.segments.append(current_segment)\n",
        "    return Genomes\n",
        "\n",
        "def get_clusters_of_length_d_at_least_q_times(Genomes, d, q):\n",
        "    segment_counts = {}\n",
        "\n",
        "    for Genome_id, Genome in Genomes.items():\n",
        "        for segment in Genome.segments:\n",
        "            if len(segment) >= d:\n",
        "                for i in range(len(segment) - d + 1):\n",
        "                    sub_segment = segment[i:i+d]\n",
        "                    segment_tuple = tuple(sorted(sub_segment))\n",
        "                    if segment_tuple in segment_counts:\n",
        "                        if Genome_id not in segment_counts[segment_tuple]:\n",
        "                            segment_counts[segment_tuple][Genome_id] = (sub_segment, Genome.creature_name)\n",
        "                    else:\n",
        "                        segment_counts[segment_tuple] = {Genome_id: (sub_segment, Genome.creature_name)}\n",
        "\n",
        "    filtered_clusters = {}\n",
        "    for segment, genomes in segment_counts.items():\n",
        "        if len(genomes) >= q:\n",
        "            filtered_clusters[segment] = genomes\n",
        "    top_clusters = dict(sorted(filtered_clusters.items(), key=lambda item: len(item[1]), reverse=True)[:5])\n",
        "    return top_clusters\n",
        "\n",
        "# Main code\n",
        "q = 5\n",
        "d = 12\n",
        "while (d<14):\n",
        "  filename = 'new_cog_words_bac.txt'\n",
        "  start_time = datetime.now()\n",
        "  Genomes = process_file(filename)\n",
        "\n",
        "  selected_lists = get_clusters_of_length_d_at_least_q_times(Genomes, d, q)\n",
        "  timedelta = datetime.now() - start_time\n",
        "  print(timedelta)\n",
        "  print()\n",
        "  print(\" \"*30 + \"Top 5 clusters of length\", d, \"appearing at least\", q, \"times:\")\n",
        "  print()\n",
        "  for segment, cluster_info in selected_lists.items():\n",
        "      print(\"Cluster:\", segment)\n",
        "      print(\"Occurrences:\", len(cluster_info))\n",
        "      print(\"Genome ID's and Creature Names:\")\n",
        "      for genome_id, (order, creature_name) in cluster_info.items():\n",
        "          print(f\"\\tGenome ID: {genome_id}, Creature Name: {creature_name}, Order: {order}\")\n",
        "      print(\"First Instance of this Cluster:\", next(iter(cluster_info.values()))[0])\n",
        "      print(\"Information about the COG's:\")\n",
        "      for cog in segment:\n",
        "        if cog in cog_dict:\n",
        "          print(\"\\t COG\",cog, \":\", cog_dict[cog])\n",
        "        else:\n",
        "          print(\"\\t COG\",cog, \":\", \"No Information was found for this COG\")\n",
        "      print(\"Statistics about the different orders we found:\")\n",
        "      word_counter = {}\n",
        "      host_list = {}\n",
        "      for value in cluster_info.values():\n",
        "        word_string = ', '.join(value[0])\n",
        "        word_counter[word_string] = word_counter.get(word_string, 0) + 1\n",
        "        host_name = value[1]\n",
        "        if word_string in host_list:\n",
        "          host_list[word_string].append(host_name)\n",
        "        else:\n",
        "          host_list[word_string]= []\n",
        "          host_list[word_string].append(host_name)\n",
        "      for key, value in word_counter.items():\n",
        "        print(\"\\t [\", key , \"]:\", value, \"times out of\", len(cluster_info), \"which is\", round(100*int(value)/len(cluster_info),2), \"percent\")\n",
        "      print(\"Statistics about the Phylum:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Phylum:\")\n",
        "        phylum_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_phylum = taxa_dict[host][1]\n",
        "          phylum_counter[curr_phylum] = phylum_counter.get(curr_phylum, 0) + 1\n",
        "        for key, value in phylum_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print(\"Statistics about the Habitat:\")\n",
        "      for order in host_list.keys():\n",
        "        print(\"\\t Order: [\", order , \"] - Appears in these Habitats:\")\n",
        "        habitat_counter = {}\n",
        "        total_hosts = len(host_list[order])\n",
        "        for host in host_list[order]:\n",
        "          curr_habitat = habitat_dict[host]\n",
        "          habitat_counter[curr_habitat] = habitat_counter.get(curr_habitat, 0) + 1\n",
        "        for key, value in habitat_counter.items():\n",
        "          print(\"\\t\\t\", key, \"-\", value, \"times out of\", total_hosts, \",which is\", round(100*int(value)/total_hosts,2), \"percent\")\n",
        "      print()\n",
        "      print(\"*\"*100)\n",
        "  d = d+1\n",
        "  print(\"*\"*100)\n",
        "  print(\"*\"*100)"
      ],
      "metadata": {
        "id": "6gkYtkff_aR-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Comparing two Profiles of the same Cluster - Part 2\n",
        "\n",
        "> Indented block"
      ],
      "metadata": {
        "id": "AMSybJsI62h8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#functions for profile comaring\n",
        "def occurence_score(cluster,db1_clusters,db2_clusters,db1_lines,db2_lines):\n",
        "  score_chrom = len(db1_clusters[cluster])/(db1_lines)\n",
        "  score_plas = len(db2_clusters[cluster])/(db2_lines)\n",
        "  return 100-100*abs(score_chrom-score_plas)\n",
        "\n",
        "def top_three_score(set1,set2):\n",
        "  common_elements = set1.intersection(set2)\n",
        "  return ((len(common_elements)*2)/(len(set1)+len(set2)))*100\n",
        "\n",
        "def phylum_diversity_score(phl_count1,phl_count2):\n",
        "  return (1-(abs(phl_count1-phl_count2)/max(phl_count1,phl_count2)))*100\n",
        "\n",
        "\n",
        "\n",
        "def get_top_three_elements(db,db_clusters,cluster, indicator):\n",
        "    element_count = {};\n",
        "    for genome_id in db_clusters[cluster].keys():\n",
        "      curr_element = str(data_by_indicator(db,genome_id,cluster,db_clusters,indicator))\n",
        "      if curr_element in element_count and curr_element != \"-\":\n",
        "        element_count[curr_element] += 1\n",
        "      else:\n",
        "        element_count[curr_element] = 1\n",
        "    return set(sorted(element_count, key=element_count.get, reverse=True)[:3])\n",
        "\n",
        "def data_by_indicator(db,genome_id,cluster,db_clusters,indicator):\n",
        "  host_name = db[genome_id].creature_name\n",
        "  if(indicator == 2): #phylum\n",
        "    return taxa_dict[host_name][1]\n",
        "  if(indicator == 4): #hosts\n",
        "    return host_name\n",
        "  if(indicator == 5): #habitats\n",
        "    return habitat_dict[host_name]\n",
        "  if(indicator == 6): #orders\n",
        "    return db_clusters[cluster][genome_id][0]\n",
        "\n",
        "def score(cluster,db1,db2,db1_clusters,db2_clusters,db1_file,db2_file,score_normalize):\n",
        "  scores = [];\n",
        "  #get occurence score.\n",
        "  scores.append(occurence_score(cluster,db1_clusters,db2_clusters,db1_file,db2_file))\n",
        "  #get score: phylum. indicator 2 (as numbered in the description)\n",
        "  top3_phylum_db1 = get_top_three_elements(db1,db1_clusters,cluster,2)\n",
        "  top3_phylum_db2 = get_top_three_elements(db2,db2_clusters, cluster,2)\n",
        "  scores.append(top_three_score(top3_phylum_db1,top3_phylum_db2))\n",
        "  #get score: phylum diversity\n",
        "  phylum_list_db1 = [];\n",
        "  phylum_list_db2 = [];\n",
        "  for genome_id in db1_clusters[cluster].keys():\n",
        "    host_name = db1[genome_id].creature_name\n",
        "    curr_phylum = taxa_dict[host_name][1]\n",
        "    if curr_phylum not in phylum_list_db1:\n",
        "      phylum_list_db1.append(curr_phylum)\n",
        "    if curr_phylum not in phylum_list_db2:\n",
        "      phylum_list_db2.append(curr_phylum)\n",
        "  scores.append(phylum_diversity_score(len(phylum_list_db1),len(phylum_list_db2)))\n",
        "  #get score:hosts. indicator 4 (as numbered in the description)\n",
        "  top3_host_names_db1 = get_top_three_elements(db1,db1_clusters,cluster,4)\n",
        "  top3_host_names_db2 = get_top_three_elements(db2,db2_clusters, cluster,4)\n",
        "  scores.append(top_three_score(top3_host_names_db1,top3_host_names_db2))\n",
        "  #get score: habitats. indicator 5 (as numbered in the description)\n",
        "  top3_habitats_db1 = get_top_three_elements(db1,db1_clusters,cluster,5)\n",
        "  top3_habitats_db2 = get_top_three_elements(db2,db2_clusters, cluster,5)\n",
        "  scores.append(top_three_score(top3_habitats_db1,top3_habitats_db2))\n",
        "  #get score: orders. indicator 6 (as numbered in the description)\n",
        "  top3_orders_db1 = get_top_three_elements(db1,db1_clusters,cluster,6)\n",
        "  top3_orders_db2 = get_top_three_elements(db2,db2_clusters, cluster,6)\n",
        "  scores.append(top_three_score(top3_habitats_db1,top3_habitats_db2))\n",
        "  #calculate final score:\n",
        "  final_score = 0\n",
        "  for i in range(5):\n",
        "    final_score += scores[i]*score_normalize[i]\n",
        "  return final_score\n",
        "\n",
        "\n",
        "def get_all_clusters_of_length_d_at_least_q_times(Genomes, d, q):\n",
        "    segment_counts = {}\n",
        "    for Genome_id, Genome in Genomes.items():\n",
        "        for segment in Genome.segments:\n",
        "            if len(segment) >= d:\n",
        "                for i in range(len(segment) - d + 1):\n",
        "                    sub_segment = segment[i:i+d]\n",
        "                    segment_tuple = tuple(sorted(sub_segment))\n",
        "                    if segment_tuple in segment_counts:\n",
        "                        if Genome_id not in segment_counts[segment_tuple]:\n",
        "                            segment_counts[segment_tuple][Genome_id] = (sub_segment, Genome.creature_name)\n",
        "                    else:\n",
        "                        segment_counts[segment_tuple] = {Genome_id: (sub_segment, Genome.creature_name)}\n",
        "\n",
        "    filtered_clusters = {}\n",
        "    for segment, genomes in segment_counts.items():\n",
        "        if len(genomes) >= q:\n",
        "            filtered_clusters[segment] = genomes\n",
        "    return filtered_clusters\n",
        "\n"
      ],
      "metadata": {
        "id": "xsViQEEEyXe2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#data printing function:\n",
        "def print_data(db,cluster,cluster_dict):\n",
        "  print(\"Numbers of appearances: \")\n",
        "  print(len(cluster_dict[cluster].keys()))\n",
        "  print(\"Top 3 phylums:\")\n",
        "  print(get_top_three_elements(db,cluster_dict,cluster,2))\n",
        "  phylum_list = [];\n",
        "  for genome_id in cluster_dict[cluster].keys():\n",
        "    host_name = db[genome_id].creature_name\n",
        "    curr_phylum = taxa_dict[host_name][1]\n",
        "    if curr_phylum not in phylum_list:\n",
        "      phylum_list.append(curr_phylum)\n",
        "  print(\"number of phylums: \")\n",
        "  print(len(phylum_list))\n",
        "  print(\"\\n\")\n",
        "  print(\"Top 3 hosts: \")\n",
        "  print(get_top_three_elements(db,cluster_dict,cluster,4))\n",
        "  print(\"\\n\")\n",
        "  print(\"Top 3 habitat: \")\n",
        "  print( get_top_three_elements(db,cluster_dict,cluster,5))\n",
        "  print(\"\\n\")\n",
        "  print(\"Top 3 orders: \")\n",
        "  print(get_top_three_elements(db,cluster_dict,cluster,6))"
      ],
      "metadata": {
        "id": "kCPeypxHhqmv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_clusters_that_differ_significantly(db1,db2,q1,q2,d,db1_lines,db2_lines,score_normalize,bar_score):\n",
        "  db1_clusters = get_all_clusters_of_length_d_at_least_q_times(db1,d,q1)\n",
        "  db2_clusters = get_all_clusters_of_length_d_at_least_q_times(db2,d,q2)\n",
        "  #get the clusters found in both maps:\n",
        "  common_clusters = set(db1_clusters.keys()) & set(db2_clusters.keys())\n",
        "  for cluster in common_clusters:\n",
        "    curr_score = score(cluster,db1,db2,db1_clusters,db2_clusters,db1_lines,db2_lines,score_normalize)\n",
        "\n",
        "    if curr_score<bar_score:\n",
        "      return [cluster,db1_clusters,db2_clusters]\n",
        "\n",
        "  #if no fitting cluster was found. we return 0\n",
        "  return 0\n",
        "\n",
        "\n",
        "#main code:\n",
        "q1 = 316\n",
        "q2 = 54\n",
        "d = 4\n",
        "score_normalize = [0.1,0.2,0.1,0.2,0.2,0.2]\n",
        "bar_score = 40\n",
        "chromosome_filename = 'new_cog_words_bac.txt'\n",
        "plasmid_filename = 'cog_words_plasmid.txt'\n",
        "chromosome_file = open(chromosome_filename, 'r')\n",
        "plasmid_file = open(plasmid_filename, 'r')\n",
        "lines_chromosome =sum(1 for line in chromosome_file)\n",
        "lines_plasmid =sum(1 for line in plasmid_file)\n",
        "chromosome_genomes = process_file(chromosome_filename)\n",
        "plasmid_genomes = process_file(plasmid_filename)\n",
        "cluster  = get_clusters_that_differ_significantly(chromosome_genomes,plasmid_genomes,q1,q2,d,lines_chromosome,lines_plasmid,score_normalize,bar_score)\n",
        "if(cluster == 0):\n",
        "  print(\"No cluster found! change query\")\n",
        "else:\n",
        "  print(\" \"*30 + \"Cluster found: \")\n",
        "  print(\" \"*28 , cluster[0])\n",
        "  print(\"Data from \" + chromosome_filename + \": \")\n",
        "  print_data(chromosome_genomes,cluster[0],cluster[1])\n",
        "  print(\"\\n\")\n",
        "  print(\"\\n\")\n",
        "  print(\"Data from \" + plasmid_filename + \": \")\n",
        "  print_data(plasmid_genomes,cluster[0],cluster[2])"
      ],
      "metadata": {
        "id": "5S4n6F-fBl6g",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "IosWtpgWpY8-",
        "Pkpzz7eTSnio",
        "VXgh4wT76qpO"
      ],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}